### PR TITLE
Lots of improvements

### PR DESCRIPTION
**Pull request by sdroege**
This fixes various memory unsafety issues, memory leaks and also cleans up the code otherwise.

I'll do more work on the FIXME comments in the next days and also fix various other issues that exist currently.

### CODE REVIEW

**Comment by sdroege** [comment]
And generally I would always recommend to keep the unsafe code very localized in safe abstractions, and not spread out all over the codebase, or otherwise there's not much difference to writing everything in C.

**Comment by sdroege** [comment]
Lots of further updates here but I'm not done yet.

**Comment by sdroege** [comment]
Next step is to make all the waiting interruptable and make use of the NDI SDK timestamps in a meaningful way.

**Comment by sdroege** [comment]
And then proper error handling everywhere.

**Comment by sdroege** [comment]
This is now finished from my side.

**Comment by rubenrua** [comment]
I'm very pleased to review this code next week.

**Comment by androidwiltron** [comment]
which ndi SDK is this against? great work btw

**Comment by sdroege** [comment]
On 19 July 2019 21:49:21 EEST, androidwiltron <notifications@github.com> wrote:
>which ndi SDK is this against?

I don't think that I needs a newer version than before, no special new SDK API is used. I was using the latest release of the SDK as of two weeks ago.

It should work with pre-3 streams and at least any 3 version of the SDK, if not older.  I'd have to check in detail next week.

Are you running into specific problems?


-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: Cargo.toml [hunk 1]
```diff
@@ -8,16 +8,23 @@ description = "NewTek NDI Plugin"
 
 [dependencies]
 glib = { version = "0.8.0", features = ["subclassing"] }
-gstreamer = { version = "0.14.0", features = ["subclassing"] }
+gobject-sys = "0.9"
+gstreamer = { version = "0.14.3", features = ["subclassing", "v1_12"] }
 gstreamer-base = { version = "0.14.0", features = ["subclassing"] }
 gstreamer-audio = "0.14.0"
-gstreamer-video = "0.14.0"
+gstreamer-video = "0.14.3"
```
Comment (rubenrua) [inline Cargo.toml]:
```suggestion
gstreamer-video = { version = "0.14.3", features = ["v1_12"] }
```

Using ubuntu 18.04 (gstreamer version 1.14.4-1~ubuntu18.04.1) and cargo 1.36.0 (c4fcfb725 2019-05-15). 


<details>
 <summary>Error:</summary>

```
$ cargo build --no-default-features
...
   Compiling gst-plugin-ndi v1.0.0 (/home/rubenrua/src/github.com/teltek/gst-plugin-ndi)
error[E0433]: failed to resolve: could not find `VideoFieldOrder` in `gst_video`
    --> src/receiver.rs:1339:58
     |
1339 |                 builder = builder.field_order(gst_video::VideoFieldOrder::TopFieldFirst);
     |                                                          ^^^^^^^^^^^^^^^ could not find `VideoFieldOrder` in `gst_video`

error[E0599]: no method named `field_order` found for type `gst_video::VideoInfoBuilder<'_>` in the current scope
    --> src/receiver.rs:1339:35
     |
1339 |                 builder = builder.field_order(gst_video::VideoFieldOrder::TopFieldFirst);
     |                                   ^^^^^^^^^^^

```
</details>

File: src/receiver.rs [hunk 2]
```diff
@@ -0,0 +1,1769 @@
...
+            {
+                gst_element_error!(
+                    element,
+                    gst::StreamError::Format,
+                    ["Separate field interlacing not supported"]
+                );
+                return Err(gst::FlowError::NotNegotiated);
+            }
+
+            let builder = gst_video::VideoInfo::new(
```
Comment (rubenrua) [inline src/receiver.rs]:
```suggestion
            let mut builder = gst_video::VideoInfo::new(
```

Error: 

```
error[E0384]: cannot assign twice to immutable variable `builder`
    --> src/receiver.rs:1339:17
     |
1319 |             let builder = gst_video::VideoInfo::new(
     |                 -------
     |                 |
     |                 first assignment to `builder`
     |                 help: make this binding mutable: `mut builder`
...
1339 |                 builder = builder.field_order(gst_video::VideoFieldOrder::TopFieldFirst);
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot assign twice to immutable variable
```

**Comment by rubenrua** [comment]
Strange error with `cargo build --no-default-features` using ubuntu 18.04 (gstreamer version 1.14.4-1~ubuntu18.04.1) and cargo 1.36.0 (c4fcfb725 2019-05-15)  

<details>
 <summary>Error</summary>

```
   Compiling gst-plugin-ndi v1.0.0 (/home/rubenrua/src/github.com/teltek/gst-plugin-ndi)
error[E0573]: expected type, found static `TIMECODE_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0574]: expected struct, variant or union type, found static `TIMECODE_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a struct, variant or union type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0573]: expected type, found static `TIMECODE_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0573]: expected type, found static `TIMECODE_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0573]: expected type, found static `TIMESTAMP_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0574]: expected struct, variant or union type, found static `TIMESTAMP_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a struct, variant or union type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0573]: expected type, found static `TIMESTAMP_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0573]: expected type, found static `TIMESTAMP_CAPS`
  --> src/lib.rs:46:1
   |
46 | / lazy_static! {
47 | |     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
48 | |         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
49 | |     };
...  |
59 | |     };
60 | | }
   | |_^ not a type
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 8 previous errors

error: Could not compile `gst-plugin-ndi`.

To learn more, run the command again with --verbose.
```
</details>

A possible solution could be using two lazy_static blocks:

```diff
diff --git a/src/lib.rs b/src/lib.rs
index 596496e..61b5850 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -47,13 +47,14 @@ lazy_static! {
     static ref DEFAULT_RECEIVER_NDI_NAME: String = {
         format!("GStreamer NDI Source {}-{}", env!("CARGO_PKG_VERSION"), env!("COMMIT_ID"))
     };
+}

-    #[cfg(feature = "reference-timestamps")]
+#[cfg(feature = "reference-timestamps")]
+lazy_static! {
     static ref TIMECODE_CAPS: gst::Caps = {
         gst::Caps::new_simple("timestamp/x-ndi-timecode", &[])
     };

-    #[cfg(feature = "reference-timestamps")]
     static ref TIMESTAMP_CAPS: gst::Caps = {
         gst::Caps::new_simple("timestamp/x-ndi-timestamp", &[])
     };
```

**Comment by sdroege** [comment]
Thanks, apparently I didn't test without the default features in the later commits. Will do so now!

**Comment by sdroege** [comment]
Should be fine now, sorry :)

**Comment by rubenrua** [comment]
Thank you

**Pull request merged**